src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,475 @@
+				return fmt.Errorf("could not unmarshal PipelineInfo bytes from PFS: %v", err)
+			}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("sidecar s3 gateway: could not read pipeline spec commit: %v", err)
+		}
+		if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+			return nil // break out of backoff (nothing to serve via S3 gateway)
+		}
+
+		// Set auth token for s.pachClient (pipelinePtr.AuthToken will be empty if
+		// auth is off)
+		pipelineName := s.pipelineInfo.Pipeline.Name
+		pipelinePtr := &pps.EtcdPipelineInfo{}
+		err := a.pipelines.ReadOnly(retryCtx).Get(pipelineName, pipelinePtr)
+		if err != nil {
+			return fmt.Errorf("could not get auth token from etcdPipelineInfo: %v", err)
+		}
+		s.pachClient.SetAuthToken(pipelinePtr.AuthToken)
+		return nil
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		logrus.Errorf("error starting sidecar s3 gateway: %v; retrying in %d", err, d)
+		return nil
+	}); err != nil {
+		// This code should never run, but I hesitate to introduce a panic to new
+		// code
+		logrus.Errorf("restarting startup of sidecar s3 gateway: %v", err)
+		a.ServeSidecarS3G()
+	}
+	if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+		return // break early (nothing to serve via S3 gateway)
+	}
+
+	// begin creating k8s services and s3 gateway instances for each job
+	done := make(chan string)
+	go func() {
+		s.createK8sServices()
+		done <- "createK8sServices"
+	}()
+	go func() {
+		s.serveS3Instances()
+		done <- "serveS3Instances"
+	}()
+	finisher := <-done
+	panic(
+		fmt.Sprintf("sidecar s3 gateway: %s is exiting, which should never happen", finisher),
+	)
+}
+
+type jobHandler struct {
+	s3g *sidecarS3G
+
+	// OnCreate runs when a job is created. Should be idempotent
+	OnCreate func(ctx context.Context, jobInfo *pps.JobInfo)
+
+	// OnTerminate runs when a job ends. Should be idempotent
+	OnTerminate func(ctx context.Context, jobID string)
+}
+
+func (s *sidecarS3G) serveS3Instances() {
+	// Watch for new jobs & initialize s3g for each new job
+	NewS3InstanceCreatingJobHandler(s).start()
+	panic("sidecar s3 gateway: serveS3Instances() is exiting, which shouldn't happen")
+}
+
+func (s *sidecarS3G) createK8sServices() {
+	logrus.Infof("Launching sidecar s3 gateway master process")
+	// createK8sServices goes through master election so that only one k8s service
+	// is created per pachyderm job running sidecar s3 gateway
+	backoff.RetryNotify(func() error {
+		masterLock := dlock.NewDLock(s.apiServer.env.GetEtcdClient(),
+			path.Join(s.apiServer.etcdPrefix,
+				s3gSidecarLockPath,
+				s.pipelineInfo.Pipeline.Name,
+				s.pipelineInfo.Salt))
+		ctx, err := masterLock.Lock(s.pachClient.Ctx())
+		if err != nil {
+			// retry obtaining lock
+			return fmt.Errorf("error obtaining mastership: %v", err)
+		}
+
+		// Watch for new jobs & create kubernetes service for each new job
+		NewK8SServiceCreatingJobHandler(s).start()
```

Comment:
is there any chance of this panicking, causing the dlock to be held?

Reply:
DLock expires 15 seconds after the last keepalive that it receives, so if the node dies another node will takeover mastership 15 seconds later: https://github.com/pachyderm/pachyderm/blob/master/src/server/pkg/dlock/dlock.go#L42

(if that weren't possible then master election would be quite brittle indeed :joy:)